github-runner-manager/src/github_runner_manager/platform/github_provider.py

```diff
@@ -143,17 +164,61 @@ def get_runners_health(self, requested_runners: list[RunnerIdentity]) -> Runners
             non_requested_runners=non_requested_runners,
         )
 
-    def delete_runner(self, runner_identity: RunnerIdentity) -> None:
-        """Delete a runner from GitHub.
+    def delete_runners(self, runner_ids: list[str], platform: str = "github") -> list[str]:
+        """Delete runners from GitHub.
 
-        This method will raise DeleteRunnerBusyError if the runner is not deletable, that is,
-        if it is busy. If the runner does not exist it will not fail.
+        This method will ignore DeleteRunnerBusyErrors and print a warning log.
 
         Args:
-            runner_identity: Identity of the runner to delete.
+            runner_ids: The GitHub runner IDs to delete.
+            platform: TODO: Unused argument due to the poor architecture of the provider
+                classes. The multiplexer provider should be a wrapper around the platforms, not on
+                the same level.
+
+        Returns:
+            The runner IDs that were deleted successfully.
         """
-        logger.info("Delete runner in GitHub: %s", runner_identity)
-        self._client.delete_runner(self._path, int(runner_identity.metadata.runner_id))
+        logger.info("Delete runners from GitHub provider: %s", runner_ids)
+        # Guard multiprocessing.Pool from having 0 processes which will raise an error.
+        if not runner_ids:
+            return []
+
+        delete_configs = [
+            _DeleteRunnerConfig(runner_id=runner_id, path=self._path, github_client=self._client)
+            for runner_id in runner_ids
+        ]
+        deleted_runner_ids: list[str] = []
+        with multiprocessing.Pool(min(len(runner_ids), 30)) as pool:
+            for deleted_runner_id in pool.imap_unordered(
+                GitHubRunnerPlatform._delete_runner, delete_configs
+            ):
+                if not deleted_runner_id:
+                    continue
+                deleted_runner_ids.append(deleted_runner_id)
+        return deleted_runner_ids
+
+    @staticmethod
+    def _delete_runner(delete_runner_config: _DeleteRunnerConfig) -> str | None:
+        """Delete a single runner from GitHub.
+
+        This method is a wrapper to be called via multiprocessing pool for parallel deletion.
+
+        Args:
+            delete_runner_config: The configuration to use for deleting the runner.
+
+        Returns:
+            The runner ID of the deleted runner
+        """
+        try:
+            delete_runner_config.github_client.delete_runner(
+                path=delete_runner_config.path, runner_id=int(delete_runner_config.runner_id)
+            )
+        except DeleteRunnerBusyError:
+            logger.warning(
+                "Delete runner attempt failed, busy runner: %s", delete_runner_config.runner_id
+            )
+            return None
```

Comment:
Instead of catching here can we catch it in the caller to avoid returning None and have that failure explicit in the caller method?

Reply:
One advantage of keeping the concurrency at the lower levels (GitHub API wrapper, OpenStack cloud API wrapper) is that we can easily share connections if required.
One advantage of keeping the concurrency at higher levels (GitHub runner manager, OpenStack cloud manager) is that we can be transparent about who's doing concurrency.
Happy to implement either ways.

Reply:
Ah sorry, the GitHub provider is fine, but the `openstack_cloud.py` which is essentially an API wrapper class for OpenStack has `delete_instances` which does the multi-threading. I would like to move this up one level to the `OpenStackRunnerManager` class so that the multithreading is transparent and the API is also transparent about the capabilities (no support for batch delete).